src/nbl/video/CVKLogicalDevice.h

```diff
@@ -124,51 +327,472 @@ class CVKLogicalDevice final : public ILogicalDevice
+            bufferMemoryReqs.sharingMode = initialMreqs.sharingMode;
+            bufferMemoryReqs.queueFamilyIndices = initialMreqs.queueFamilyIndices;
+
+            // Not sure if I'd actually need these
+            bufferMemoryReqs.memoryHeapLocation = initialMreqs.memoryHeapLocation;
+            bufferMemoryReqs.mappingCapability = initialMreqs.mappingCapability;
+            bufferMemoryReqs.prefersDedicatedAllocation = initialMreqs.prefersDedicatedAllocation;
+            bufferMemoryReqs.requiresDedicatedAllocation = initialMreqs.requiresDedicatedAllocation;
+
+            vkGetBufferMemoryRequirements(m_vkdev, vk_buffer, &bufferMemoryReqs.vulkanReqs);
+
+            return core::make_smart_refctd_ptr<CVulkanBuffer>(
+                core::smart_refctd_ptr<CVKLogicalDevice>(this), bufferMemoryReqs, vk_buffer);
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+
+    core::smart_refctd_ptr<IGPUBuffer> createGPUBufferOnDedMem(const IDriverMemoryBacked::SDriverMemoryRequirements& initialMreqs, const bool canModifySubData = false) override
+    {
+        core::smart_refctd_ptr<IGPUBuffer> gpuBuffer = createGPUBuffer(initialMreqs, false);
+
+        if (!gpuBuffer)
+            return nullptr;
+
+        // Todo(achal): Probably do not call getMemoryReqs at all but
+        // vkGetBufferMemoryRequirements or equivalent
+        core::smart_refctd_ptr<video::IDriverMemoryAllocation> bufferMemory =
+            allocateDeviceLocalMemory(gpuBuffer->getMemoryReqs());
+
+        if (!bufferMemory)
+            return nullptr;
+
+        ILogicalDevice::SBindBufferMemoryInfo bindBufferInfo = {};
+        bindBufferInfo.buffer = gpuBuffer.get();
+        bindBufferInfo.memory = bufferMemory.get();
+        bindBufferInfo.offset = 0ull;
+        bindBufferMemory(1u, &bindBufferInfo);
+
+        return gpuBuffer;
+    }
+        
     core::smart_refctd_ptr<IGPUShader> createGPUShader(core::smart_refctd_ptr<asset::ICPUShader>&& cpushader) override
     {
-        return nullptr;
+        const asset::ICPUBuffer* source = cpushader->getSPVorGLSL();
+        core::smart_refctd_ptr<asset::ICPUBuffer> clone =
+            core::smart_refctd_ptr_static_cast<asset::ICPUBuffer>(source->clone(1u));
+        if (cpushader->containsGLSL())
+        {
+            return core::make_smart_refctd_ptr<CVulkanShader>(
+                core::smart_refctd_ptr<CVKLogicalDevice>(this), std::move(clone),
+                IGPUShader::buffer_contains_glsl);
+        }
+        else
+        {
+            return core::make_smart_refctd_ptr<CVulkanShader>(core::smart_refctd_ptr<CVKLogicalDevice>(this),
+                std::move(clone));
+        }
+    }
+
+    core::smart_refctd_ptr<IGPUImage> createGPUImage(asset::IImage::SCreationParams&& params) override
+    {
+        VkImageCreateInfo vk_createInfo = { VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO };
+        vk_createInfo.pNext = nullptr; // there are a lot of extensions
```

Comment:
any of them interesting?

Reply:
> Each pNext member of any structure (including this one) in the pNext chain must be either NULL or a pointer to a valid instance of VkBufferCollectionImageCreateInfoFUCHSIA, VkDedicatedAllocationImageCreateInfoNV, VkExternalFormatANDROID, VkExternalMemoryImageCreateInfo, VkExternalMemoryImageCreateInfoNV, VkImageDrmFormatModifierExplicitCreateInfoEXT, VkImageDrmFormatModifierListCreateInfoEXT, VkImageFormatListCreateInfo, VkImageStencilUsageCreateInfo, VkImageSwapchainCreateInfoKHR, VkVideoProfileKHR, or VkVideoProfilesKHR

I suggest to explore/use them as we need them

Reply:
@Erfan-Ahmadi explore 
```
VkDedicatedAllocationImageCreateInfoNV
VkExternalFormatANDROID
VkImageFormatListCreateInfo
VkImageStencilUsageCreateInfo
VkImageStencilUsageCreateInfo
VkImageSwapchainCreateInfoKHR
```